pkg/image/fetcher_test.go

```diff
@@ -404,4 +403,47 @@ func testFetcher(t *testing.T, when spec.G, it spec.S) {
 			})
 		})
 	})
+
+	when("#CheckReadAccessValidator", func() {
+		var daemon bool
+
+		when("Daemon is true", func() {
+			it.Before(func() {
+				daemon = true
+			})
+
+			it("read access is always valid", func() {
```

Comment:
I think it's actually a bit more complicated - it is also dependent on the pull policy. If the pull policy is "always" then it's still essentially a registry operation; conversely if the pull policy is "never" then we could check if the image is already in the daemon; "if not present" is the special case where we fall back to the registry. This logic is already contained (and can be extracted from) [here](https://github.com/buildpacks/pack/blob/6a9e3d193abca7fff82ce2df10d181c63fbde832/pkg/image/fetcher.go#L100-L109).

Reply:
I am not totally sure about it, we will always have access to read from the daemon but it is possible that the image is not available there, but for me, that's a different thing. For example: 

The error reported by Sophie in this issue is:

```bash
 CheckReadAccess failed for the run image stack-run-845695c9-5147-42a2-8cb3-c0ea0500c6a4:latest, error: HEAD https://index.docker.io/v2/library/stack-run-845695c9-5147-42a2-8cb3-c0ea0500c6a4/manifests/latest: unexpected status code 401 Unauthorized (HEAD responses have no body, use GET for details)
```

- run image ref: stack-run-845695c9-5147-42a2-8cb3-c0ea0500c6a4
- Daemon: true
- Pull-Policy: "always" (default one)

according to your statement
> it is also dependent on the pull policy. If the pull policy is "always" then it's still essentially a registry operation

`CheckReadAccessValidator` should return:
-  **false**? because it must try to check read access against a registry that doesn't exist? 
-  **true**? because although it can't find the image a remote registry the image is available in the daemon? 

My confusion is: if we want `CheckReadAccessValidator` success because I do have read access from the source or success because the image exists in the source?

Reply:
I [changed](https://github.com/buildpacks/pack/pull/2092/commits/2d8f1bb15626e578289e6edf1a718e75d2dfa871#diff-76bea0285076919ef2d39e8bbb47824f3f7c991ff0a04302eb496f244f3380dc) the implementation, when `opts.Daemon` is true, we will try to check if the image is available in the docker daemon and if **it's not** we will try to check the remote only when Pull is not **Never**. I think this should cover the cases.